### ISSUE TITLE
Can't run Java 8 on Windows 10 x64 without Windows SDK?

### ISSUE THREAD

**Issue by TheLQ** [libgdx/packr#82]
I have a near vanilla Windows 10 VM that cannot run any of the executables with Java 8u111. Running with the console shows

```
Loading JVM runtime library ...
Error code [126]: The specified module could not be found.
Error: failed to load VM runtime library!
Press ENTER key to exit.
```

Things I've tried

 - Verified I'm using the `windows64` platform with 64-bit JREs
 - Tried both Oracle JRE and OpenJDK
 - Verified `jre\bin\server\jvm.dll` exists
 - Installed VC 2013 and 2015 x64 Redistributable (reset to previous snapshot between tries)
 - Searched packr documentation for any Windows runtime dependencies 
 - Watched Process Monitor although I'm not sure what I should be looking for
 - No minimize and soft minimize
 
The only thing that did work is installing the Windows SDK as specified in #49 . However I clearly can't install that on every users machine. What exactly is required to be installed on a vanilla Windows machine to run the executable?

Config below
```
{
    "platform": "windows64",
    "jdk": "/home/leon/jre-extract/java-1.8.0-openjdk-1.8.0.111-1.b15.ojdkbuild.windows.x86_64.zip",
    "executable": "brailleblaster-packr",
    "classpath": [
        "brailleblaster.jar"
    ],
    "mainclass": "org.brailleblaster.Main",
    "output": "out-packr-win"
}
```

**Comment by code-disaster** [comment]
Can you try running [Dependency Walker](http://www.dependencywalker.com/) on `jre\bin\server\jvm.dll` ? I wonder if it tells you which dependency is missing.

**Comment by TheLQ** [comment]
Here's the text file output

`````
[ ? ]  API-MS-WIN-CORE-APIQUERY-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-APPCOMPAT-L1-1-1.DLL                          Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-COMM-L1-1-0.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-CONSOLE-L1-1-0.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-CONSOLE-L2-1-0.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-DATETIME-L1-1-1.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-DATETIME-L1-1-2.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-DEBUG-L1-1-1.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-DELAYLOAD-L1-1-1.DLL                          Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-ENCLAVE-L1-1-0.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-ERRORHANDLING-L1-1-1.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-ERRORHANDLING-L1-1-3.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FIBERS-L1-1-1.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FIBERS-L2-1-1.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FILE-L1-2-1.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FILE-L1-2-2.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FILE-L2-1-1.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-FILE-L2-1-2.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-HANDLE-L1-1-0.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-HEAP-L1-2-0.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-HEAP-L2-1-0.DLL                               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-IO-L1-1-1.DLL                                 Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-JOB-L1-1-0.DLL                                Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LIBRARYLOADER-L1-2-0.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LIBRARYLOADER-L1-2-2.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LIBRARYLOADER-L2-1-0.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LOCALIZATION-L1-2-1.DLL                       Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LOCALIZATION-L2-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-LOCALIZATION-PRIVATE-L1-1-0.DLL               Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-MEMORY-L1-1-2.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-NAMEDPIPE-L1-2-0.DLL                          Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-NAMEDPIPE-L1-2-2.DLL                          Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-NAMESPACE-L1-1-0.DLL                          Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-NORMALIZATION-L1-1-0.DLL                      Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROCESSENVIRONMENT-L1-2-0.DLL                 Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROCESSSNAPSHOT-L1-1-0.DLL                    Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROCESSTHREADS-L1-1-2.DLL                     Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROCESSTHREADS-L1-1-3.DLL                     Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROCESSTOPOLOGY-L1-2-0.DLL                    Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PROFILE-L1-1-0.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PSAPI-ANSI-L1-1-0.DLL                         Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-PSAPI-L1-1-0.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-REALTIME-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-REGISTRY-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-RTLSUPPORT-L1-2-0.DLL                         Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-SIDEBYSIDE-L1-1-0.DLL                         Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-STRING-L1-1-0.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-SYNCH-L1-2-0.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-SYNCH-L1-2-1.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-SYSINFO-L1-2-1.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-SYSINFO-L1-2-3.DLL                            Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-THREADPOOL-LEGACY-L1-1-0.DLL                  Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-THREADPOOL-PRIVATE-L1-1-0.DLL                 Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-TIMEZONE-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-WINDOWSERRORREPORTING-L1-1-0.DLL              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-WOW64-L1-1-0.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-WOW64-L1-1-1.DLL                              Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-CORE-XSTATE-L2-1-0.DLL                             Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-EVENTING-PROVIDER-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-SECURITY-APPCONTAINER-L1-1-0.DLL                   Error opening file. The system cannot find the file specified (2).
[ ? ]  API-MS-WIN-SECURITY-BASE-L1-2-0.DLL                           Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-ADVAPI32-NPUSERNAME-L1-1-0.DLL                     Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-ADVAPI32-NTMARTA-L1-1-0.DLL                        Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-ADVAPI32-REGISTRY-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-APPMODEL-DAXCORE-L1-1-0.DLL                        Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-APPXDEPLOYMENTCLIENT-APPXDEPLOY-L1-1-0.DLL         Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-APPXDEPLOYMENTCLIENT-APPXDEPLOYONECORE-L1-1-0.DLL  Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-COM-PSMREGISTER-L1-2-1.DLL                         Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-FECLIENT-ENCRYPTEDFILE-L1-1-0.DLL                  Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-GPAPI-GROUPPOLICY-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-APPCOMPAT-L1-1-0.DLL                      Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-DATETIME-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-ERRORHANDLING-L1-1-0.DLL                  Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-FILE-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-QUIRKS-L1-1-0.DLL                         Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-QUIRKS-L1-1-1.DLL                         Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-REGISTRY-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-SIDEBYSIDE-L1-1-0.DLL                     Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-WINDOWSERRORREPORTING-L1-1-0.DLL          Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNEL32-WINDOWSERRORREPORTING-L1-1-1.DLL          Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-KERNELBASE-PROCESSTHREAD-L1-1-0.DLL                Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-MRMCORER-RESMANAGER-L1-1-0.DLL                     Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-NTDSAPI-ACTIVEDIRECTORYCLIENT-L1-1-0.DLL           Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-NTDSAPI-ACTIVEDIRECTORYCLIENT-L1-1-1.DLL           Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-NTUSER-STRING-L1-1-0.DLL                           Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-SECURITY-CAPAUTHZ-L1-1-0.DLL                       Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-SECURITY-EFSWRT-L1-1-1.DLL                         Error opening file. The system cannot find the file specified (2).
[D? ]  EXT-MS-WIN-SHELL32-SHELLCOM-L1-1-0.DLL                        Error opening file. The system cannot find the file specified (2).
[  6]  BRAILLEBLASTER-PACKR.EXE                                      10/31/2016 10:29a  10/16/2016  8:27a    270,848         0x00000000     0x0004292E     x64  GUI        None        0x0000000140000000  Unknown      0x00049000    Not Loaded  N/A             N/A             0.0        12.0        6.0     6.0
[  6]  KERNEL32.DLL                                                  07/16/2016  4:42a  07/15/2016  7:21p    699,880  A      0x000B4BAA     0x000B4BAA     x64  Console    CV,Unknown  0x0000000180000000  Unknown      0x000AB000    Not Loaded  10.0.14393.0    10.0.14393.0    10.0       14.0        10.0    10.0
[  6]  KERNELBASE.DLL                                                10/05/2016  3:31a  10/05/2016  2:16a  2,213,248  A      0x00229AEE     0x00229AEE     x64  Console    CV,Unknown  0x0000000180000000  Unknown      0x0021D000    Not Loaded  10.0.14393.321  10.0.14393.321  10.0       14.0        10.0    10.0
[  6]  NTDLL.DLL                                                     10/14/2016  9:47p  10/14/2016  8:32p  1,883,784  A      0x001DA630     0x001DA630     x64  Console    CV,Unknown  0x0000000180000000  Unknown      0x001D1000    Not Loaded  10.0.14393.351  10.0.14393.351  10.0       14.0        10.0    10.0
``````

**Comment by TheLQ** [comment]
Lots of files are missing, see below. Note I just remade the VM as Java was installed on the old one so it represents a normal user's machine. Both OpenJDK and Oracle JRE can run my app if called manually. 

[jvm.dll.openjdk.txt](https://github.com/libgdx/packr/files/564098/jvm.dll.openjdk.txt)
[brailleblaster-packr.txt](https://github.com/libgdx/packr/files/564118/brailleblaster-packr.txt)

**Comment by code-disaster** [comment]
Your log reveals msvcr100.dll is missing on the clean install. It works when running the JRE manually because this DLL is included in jre/bin. If you don't want users to install the VC runtime redistributables (the full Windows SDK shouldn't be needed), it may be sufficient to pick the DLL from jre/bin and put it right next to the packr executable.

**Comment by kcpeppe** [comment]
This is a crazy work-around. Is there any way to correct the path/ld_library_path so that the proper dll's are picked up? I'm concerned that this might cause some difficulties when distributing to other Window distro's.

**Comment by code-disaster** [comment]
Windows DLL dependency hell. At least msvcr100.dll is the only non-system DLL required, and it only depends on kernel32.dll by itself. Also I'm not sure if this is true for all JRE variants out there, though it seems to be the case for the Oracle and OpenJDK versions I looked at.

This is why you can't install almost any game on Windows without its installer sneaking in one of the VC runtime distributables.

One idea I had is to add a fallback, which, when detecting a DLL load error, tries to load msvcr100.dll first, and then retries to load jvm.dll. Another (brute-force) option would be to let packr just copy the DLL during build creation.

**Comment by kcpeppe** [comment]
Wow, you'd think that after more than 25 years this problem would be solved.

**Comment by code-disaster** [comment]
I don't have a pristine Windows system to test with, so it'd be cool if one of you can try to replace your 32-bit or 64-bit executable with the ones found here:

https://github.com/code-disaster/packr/tree/%2382/src/main/resources

**Comment by kcpeppe** [comment]
My pristine install of 10 yields the following error message.

The program or feature "......\packr\output\myexec.exe" cannot start or run due to incompatibility with 64-bit versions of Windows. Please contact the software vendor to ask if a 64-bit Windows compatible version is available.

**Comment by code-disaster** [comment]
If this is a 64-bit version you are building (so you bundle a 64-bit JRE too), you only need to copy/replace packr-windows-x64.exe. The 32-bit executable won't work here.

**Comment by kcpeppe** [comment]
I replaced both.

As a follow up, I've just cloned your repository, executed mvn clean package and then packaged my app. The run produced an error code of 126. failed to load VM runtime library.

**Comment by code-disaster** [comment]
Odd. Did you checkout my "#82" branch?

**Comment by kcpeppe** [comment]
My mistake, I built again master and not "#82" branch. I've corrected and your branch solves my problem. Test application runs as expected.

**Comment by code-disaster** [comment]
Nice! Thanks a lot for testing.

**Comment by dmvanderuit** [comment]
Hi,

I just spent a day being frustrated over this issue. I successfully packaged a Java app for MacOS, but when I decided to try and make an executable for Windows (on my mac), I got the following issue:
<img width="825" alt="image" src="https://user-images.githubusercontent.com/42491822/158605418-3b3e7e7c-a3a1-47e7-837b-1e63a8be8aae.png">

For anyone else running into this issue; I accidentally bundled my `$JAVA_HOME` into the Windows-application, which, of course, is not something Windows can understand. The error message was valid, because there's no `jvm.dll` in the unix-version of the JVM. 

TLDR:
Make sure you reference the correct JDK for the OS you're building for in the `--jdk`-option when running packr.

**Comment by kidmtt94** [comment]
> Windows DLL dependency hell. At least msvcr100.dll is the only non-system DLL required, and it only depends on kernel32.dll by itself. Also I'm not sure if this is true for all JRE variants out there, though it seems to be the case for the Oracle and OpenJDK versions I looked at.
> 
> This is why you can't install almost any game on Windows without its installer sneaking in one of the VC runtime distributables.
> 
> One idea I had is to add a fallback, which, when detecting a DLL load error, tries to load msvcr100.dll first, and then retries to load jvm.dll. Another (brute-force) option would be to let packr just copy the DLL during build creation.

Hi sir, 
Does this still work in latest java 8 U351 ?
